### PR TITLE
feat(options): support whitelisting

### PR DESCRIPTION
**Pull request by Lianathanoj**
**What**: Add functionality to whitelist specified keys for sanitization.

**Why**: There may be instances where some endpoints only require certain fields to be sanitized. The default behavior before was to sanitize the entire payload for an endpoint, but now it is possible for users to choose which attributes they want sanitized via the whitelist.

**Details**:
* The whitelist only applies to the top-level attributes. If we had a payload of 
```
{
      key1: {
        nested1: 'b\0ar',
        nested2: '',
        nested3: 'word'
      },
      key2: 'bye',
      key3: {
        test: ''
      },
      key4: '\0bye '
}
```

then it is only currently possible to whitelist the top-level keys, i.e. `key1`, `key2`, `key3`, and `key4`. This means if we whitelist `key1` for the above example, then its attributes `nested1`, `nested2`, and `nested3` would all be sanitized accordingly. If we try whitelisting `nested1`, then nothing will actually be whitelisted as there is no functionality for whitelisting individual, nested attributes.

### CODE REVIEW

File: lib/sanitize.js [hunk 1]
```diff
@@ -25,25 +27,27 @@ function Sanitize (object, options) {
...
   Object.keys(object).forEach((key) => {
-    let value = object[key];
-
-    if (IsPlainObject(value)) {
-      object[key] = Sanitize(value, options);
-    } else {
-      if (IsString(value)) {
-        object[key] = value = stripNullTerminator(value).trim();
-      }
+    if (!options.whitelist || options.whitelist && Includes(options.whitelistValues, key)) {
```
Comment (sjl2) [inline lib/sanitize.js]: nit: This is probably easier to read in the inverse case:

```js
if (options.whiteList && !Includes(options.whitelistValues, key) {
  return;
}
```

Comment (sjl2) [reply lib/sanitize.js]: I'm thinking we may want to use a blacklist instead. The way we have it, we have to specify ALL of the fields we want sanitized (including nested fields).

File: README.md [hunk 2]
```diff
@@ -32,6 +32,8 @@ server.register([
   - `'replace'` - the key will be preserved, but its value will be replaced with the value of `replaceValue`.
 - `replaceValue` - valid only when `pruneMethod` is set to `'replace'`, this value will be used as the replacement of any pruned values (ie. if configured as `null`, then `{ a: '', b: 'b' }` :arrow_right: `{ a: null, b: 'b' }`).
 - `stripNull` - a boolean value to signify whether or not `null` properties should be pruned with the same `pruneMethod` and `replaceValue` as above. Defaults to `false`.
+- `whitelist` - a boolean value to signify whether or not certain properties should be whitelisted for sanitization. Defaults to `false`.
```
Comment (dmlittle) [inline README.md]: I think we can rename `whitelistValues` to just be `whitelist` and assume we apply the whitelist if it exists for that specific route. I don't think it makes sense to have `whitelistValues` but have `whitelist` be `false`.

File: lib/sanitize.js (hunk 1 above)
Comment (dmlittle) [reply lib/sanitize.js]: With a blacklist we would have the same problem where we will have to blacklist every new payload parameter, also including nested fields, that we are adding in the future. I'd rather be explicit and name the parameters we want to sanitize than the other way around.

I think we can include some sort of syntax we go by to whitelist values. For example `template` will sanitize the `template` parameter while `merge_variables.*` will whitelist everything in the `merge_variables` object. This allows us to be specific for what objects we want to sanitize but the flexibility to allow entire nested objects.

Comment (sjl2) [reply lib/sanitize.js]: But we could also do the same thing with blacklist where then we only have to say don't sanitize `merge_variables.*`. Then we only have to configure an additional thing to our endpoints.

Comment (sjl2) [reply lib/sanitize.js]: Also, I'm not even sure we'd need wildcards for this case. The idea then could be just don't sanitize any the field or nested field of blacklisted fields.

Comment (Lianathanoj) [reply lib/sanitize.js]: I think the problem scope is actually slightly different than I originally thought. There is a case where we want to actually sanitize all of the payload parameters, but we want to sanitize *differently* for certain parameters, i.e. we whitelist certain parameters to be sanitized differently while all other parameters are sanitized the same way.

The problem that this PR is currently addressing is sanitizing only certain attributes on the whitelist but not sanitizing anything else, but I'll try to address the problem above as well.

**Comment by Lianathanoj** [comment]
Closing this PR and opening up a separate PR for field overriding as overriding is more extensible than whitelisting/blacklisting.

**Pull request closed without merging**